server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateConditionsUpdater.java

```diff
@@ -0,0 +1,192 @@
+    Errors errors = new Errors();
+    validateMetric(metric, errors);
+    checkOperator(metric, operator, errors);
+    checkThresholds(warningThreshold, errorThreshold, errors);
+    checkPeriod(metric, period, errors);
+    if (!errors.isEmpty()) {
+      throw new BadRequestException(errors);
+    }
+  }
+
+  private static void validateMetric(MetricDto metric, Errors errors) {
+    errors.check(isAlertable(metric), format("Metric '%s' cannot be used to define a condition.", metric.getKey()));
+  }
+
+  private static boolean isAlertable(MetricDto metric) {
+    return isAvailableForInit(metric) && BooleanUtils.isFalse(metric.isHidden());
+  }
+
+  private static boolean isAvailableForInit(MetricDto metric) {
+    return !metric.isDataType() && !CoreMetrics.ALERT_STATUS_KEY.equals(metric.getKey());
+  }
+
+  private static void checkOperator(MetricDto metric, String operator, Errors errors) {
+    Metric.ValueType valueType = Metric.ValueType.valueOf(metric.getValueType());
+    errors.check(isOperatorAllowed(operator, valueType), format("Operator %s is not allowed for metric type %s.", operator, metric.getValueType()));
+  }
+
+  private static void checkThresholds(@Nullable String warningThreshold, @Nullable String errorThreshold, Errors errors) {
+    errors.check(warningThreshold != null || errorThreshold != null, "At least one threshold (warning, error) must be set.");
+  }
+
+  private static void checkPeriod(MetricDto metric, @Nullable Integer period, Errors errors) {
+    if (period == null) {
+      errors.check(!metric.getKey().startsWith("new_"), "A period must be selected for differential metrics.");
+    } else {
+      errors.check(period == 1, "The only valid quality gate period is 1, the leak period.");
+    }
+  }
+
+  private static void checkConditionDoesNotAlreadyExistOnSameMetricAndPeriod(Collection<QualityGateConditionDto> conditions, MetricDto metric, @Nullable final Integer period) {
+    if (conditions.isEmpty()) {
+      return;
+    }
+
+    boolean conditionExists = conditions.stream().anyMatch(new MatchMetricAndPeriod(metric.getId(), period)::apply);
+    if (conditionExists) {
+      String errorMessage = period == null
+        ? format("Condition on metric '%s' already exists.", metric.getShortName())
+        : format("Condition on metric '%s' over leak period already exists.", metric.getShortName());
+      throw new BadRequestException(errorMessage);
+    }
+  }
+
+  private static class MatchMetricAndPeriod implements Predicate<QualityGateConditionDto> {
+    private final int metricId;
+    @CheckForNull
+    private final Integer period;
+
+    public MatchMetricAndPeriod(int metricId, @Nullable Integer period) {
+      this.metricId = metricId;
+      this.period = period;
+    }
+
+    @Override
+    public boolean apply(@Nonnull QualityGateConditionDto input) {
+      return input.getMetricId() == metricId &&
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)